### PR TITLE
patterns: reject raw pointers that are not just integers

### LINKED ISSUE #116929: We're not warning against fn ptr and wide raw ptr nested inside ADTs
I found a gap in our match forward-compatibility warnings:
```rust
type T = (*const [i32], i32);
const C: T = (unsafe { std::mem::transmute((0usize, 0usize)) }, 0);

fn test(x: T) -> bool {
    matches!(x, C)
}
```
This should warn but doesn't.

The reason is that [here](https://github.com/rust-lang/rust/blob/223674a824317284b42ba197be43fd68b957ab58/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L239-L252) we are only checking the top-level type to be a wide raw ptr or fn ptr, we are not recursing.

The fix IMO is to reject those types in `compiler/rustc_trait_selection/src/traits/structural_match.rs`, which already recurses the type.

Cc @oli-obk @lcnr

### PR DESCRIPTION
**Pull request by RalfJung**
Matching against `0 as *const i32` is fine, matching against `&42 as *const i32` is not.

This extends the existing check against function pointers and wide pointers: we now uniformly reject all these pointer types during valtree construction, and then later lint because of that. See [here](https://github.com/rust-lang/rust/pull/116930#issuecomment-1784654073) for some more explanation and context.

Also fixes https://github.com/rust-lang/rust/issues/116929.

Cc @oli-obk @lcnr

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -2217,15 +2217,16 @@ declare_lint! {
...
     /// While these work in many cases as expected by users, it is possible that due to
     /// optimizations pointers are "not equal to themselves" or pointers to different functions
     /// compare as equal during runtime. This is because LLVM optimizations can deduplicate
     /// functions if their bodies are the same, thus also making pointers to these functions point
     /// to the same location. Additionally functions may get duplicated if they are instantiated
-    /// in different crates and not deduplicated again via LTO.
+    /// in different crates and not deduplicated again via LTO. Pointer identity for memory
+    /// created by `const` is similarly unreliable.
     pub POINTER_STRUCTURAL_MATCH,
-    Allow,
```
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:2228]: I didn't realize this lint is still allow-by-default oO. Seems high time we make it warn-by-default?

**Comment by RalfJung** [comment]
The 2nd commit makes this forbid-by-default so we can crater it.
@bors try

**Comment by RalfJung** [comment]
Due to https://github.com/rust-lang/rust/issues/116929 this lint still has a big gap, it'll only find bad raw pointers "at the root".

I wonder if with this change to valtree construction, we can make it so that some lint *always* fires when a const is used as a pattern and valtree construction fails? I do think those cases should all become hard errors eventually...

**Comment by RalfJung** [comment]
Let's see how much fallout we're getting from this weak form of the lint.
@craterbot check

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 2]
```diff
@@ -238,7 +240,9 @@ impl<'tcx> ConstToPat<'tcx> {
                 }
             } else if !self.saw_const_match_lint.get() {
                 match cv.ty().kind() {
-                    ty::RawPtr(pointee) if pointee.ty.is_sized(self.tcx(), self.param_env) => {}
+                    ty::RawPtr(..) if have_valtree => {
+                        // This is a good raw pointer, it was accepted by valtree construction.
+                    }
                     ty::FnPtr(..) | ty::RawPtr(..) => {
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:246]: The actual logic we want here might be "if we don't have a valtree and there was no structural match violation detected with the type-based traversal, then it must be a raw pointer thing so show the lint". That might fix https://github.com/rust-lang/rust/issues/116929. But I'm not sure if pointers are the only case where we fail to build a valtree but do not detect a structural match violation.

**Comment by RalfJung** [comment]
Lol, the doctest for `StructuralEq` uses a function pointer as a bad example.^^

File: library/core/src/marker.rs [hunk 3]
```diff
@@ -247,6 +247,7 @@ marker_impls! {
 ///
 /// const CFN: Wrap<fn(&())> = Wrap(higher_order);
 ///
+/// #[allow(pointer_structural_match)]
```
Comment (RalfJung) [inline library/core/src/marker.rs:250]: FWIW these entire trait docs are outdated; that's tracked in https://github.com/rust-lang/rust/issues/115881.

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-116930/retry-regressed-list.txt

**Comment by RalfJung** [comment]
@craterbot p=1

**Comment by RalfJung** [comment]
Argh, most of those (based on a few samples) are "deny(future_incompatible) incompatible with previous forbid". Seems like setting the default level is not a good short-cut for making this a hard error...

**Comment by RalfJung** [comment]
@craterbot check

**Comment by RalfJung** [comment]
Hm... actually it should be enough to rerun the ones we had last time.
@craterbot abort

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-116930/retry-regressed-list.txt p=1

**Comment by RalfJung** [comment]
Those seem to be 3 legit regressions:
- [this](https://github.com/h33p/ctti/blob/1ae5fa0f41c0b1366089ac063735a66e596ab667/src/impls.rs#L473C45-L473C55) matched on a fn ptr (so already triggered the lint before this PR, but the lint was allow-by-default)
- same [here](https://github.com/SaadiSave/cambridge-asm/blob/587d2274ad63e4887a9c546c60d5803deed980ab/lib/src/inst.rs#L286)
- and same [here](https://github.com/cowuake/schemus/blob/687bc85609f398e8908e39edd33744ff0aa65d7a/schemus/src/core/evaluator.rs#L53)

In particular, none of them is matching on a raw pointer that was not derived from an integer. So IMO we should land this PR to make sure people don't start doing that.

**Comment by RalfJung** [comment]
@rust-lang/lang this PR does three things:
- it fixes our old lint that detects match on function pointers and wide pointers to also catch cases where those pointers are hidden inside other types
- it makes that lint warn-by-default (but doesn't yet make it show up in dependencies)
- it makes the lint also trigger when encountering a raw pointer that was not constructed via `int as *const/mut T`, but e.g. via `&something as *const T` -- those allocations don't have guarantees on their ptr equality just like function pointers and vtables, so we should reject them for the same reason.

We haven't reached a proper conclusion in the recent "match on const" meeting, but the general direction seems to have been towards "it should still work like a pattern" (and not like sugar for `==`), and this PR is a crucial step towards ensuring that they do indeed behave like a pattern, with structural properties and all that.

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 4]
```diff
@@ -239,19 +239,15 @@ impl<'tcx> ConstToPat<'tcx> {
...
-                    ty::FnPtr(..) | ty::RawPtr(..) => {
-                        self.tcx().emit_spanned_lint(
-                            lint::builtin::POINTER_STRUCTURAL_MATCH,
-                            self.id,
-                            self.span,
-                            PointerPattern,
-                        );
-                    }
-                    _ => {}
+                if !have_valtree {
```
Comment (davidtwco) [inline compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:232]: nit: could be part of the condition on the previous line

**Review by davidtwco — APPROVED** [review]
Implementation changes look good to me, r=me after decisions from t-lang.

**Comment by nikomatsakis** [comment]
I'm in favor of this. (I also have thoughts on the semantics of matching on constants -- tl;dr I think I'm ready to accept "match is not an extensible mechanism" and hence will not act the same as `==` so we can stop talking about it)

**Comment by WaffleLapkin** [comment]
T-lang meeting consensus: let's do this, no FCP needed.

**Pull request merged**